contracts/token/CRC1155/extensions/CRC1155Enumerable.sol

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./ICRC1155Enumerable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+abstract contract CRC1155Enumerable is ERC1155, ICRC1155Enumerable {
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    // All token ids for enumeration.
+    EnumerableSet.UintSet private _allTokens;
+
+    // Mapping from token ID to token total supply.
+    mapping(uint256 => uint256) private _totalSupplies;
+
+    // Mapping from owner to owned token IDs for enumeration.
+    mapping(address => EnumerableSet.UintSet) private _ownedTokens;
+
+    /**
+     * @dev Returns the total amount of tokens stored by the contract.
+     */
+    function totalSupply() public view virtual override returns (uint256) {
+        return _allTokens.length();
+    }
+
+    /**
+     * @dev Returns the `index`-th tokenId stored by the contract.
+     */
+    function tokenByIndex(uint256 index) public view virtual override returns (uint256) {
+        return _allTokens.at(index);
+    }
```

Comment:
Btw we might consider doing a bounds check here so that users get better error messages. (Same for `tokenOfOwnerByIndex`.)

Reply:
The `EnumerableSet` already handled the boundary very well.

Reply:
Does it? It seems to me that it just accesses `set._values[index]` directly which might lead to transaction revert without any meaningful error message.

Reply:
Please refer to the OZ code [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/23869e5b2a7c6b9c3e27dee4289615b8cf50e36b/contracts/utils/structs/EnumerableSet.sol#L131)